src/Core/3DTiles/C3DTBatchTableHierarchyExtension.js

```diff
@@ -0,0 +1,149 @@
+        }
+
+        // for each instance
+        for (let i = 0; i < json.instancesLength; i++) {
+            // for each parent of the current instance
+            for (let j = 0; j < parentCounts[i]; j++) {
+                // When an instance's parentId points to itself, then it has no
+                // parent" (cf. spec)
+                if (i !== json.parentIds[parentIdsCounter]) {
+                    if (this.inverseHierarchy[i] === undefined) {
+                        this.inverseHierarchy[i] = [];
+                    }
+                    this.inverseHierarchy[i].push(json.parentIds[parentIdsCounter]);
+                    parentIdsCounter++;
+                }
+            }
+            const classId = json.classIds[i];
+            if (classCounter[classId] === undefined) {
+                classCounter[classId] = 0;
+            }
+            this.instancesIdxs[i] = {
+                classId,
+                instanceIdx: classCounter[classId],
+            };
+            classCounter[classId]++;
+        }
+    }
+
+    /**
+     * Creates and returns a javascript object holding the displayable
+     * information relative to this extension and for a given feature and
+     * its parents
+     * @param {integer} featureId - id of the feature
+     * @returns {Object} - displayable information relative to this extension
+     * and for the feature with id=featureId and for its parents
+     */
+    getPickingInfo(featureId) {
+        const instanceProperties = {};
+        // get feature class name
+        const instanceClassId = this.instancesIdxs[featureId].classId;
+        const featureClass = this.classes[instanceClassId].name;
+        // get feature properties and values
+        const instanceIdx = this.instancesIdxs[featureId].instanceIdx;
+
+        const properties = Object.keys(this.classes[instanceClassId].instances);
+        for (let i = 0; i < properties.length; i++) {
+            const property = properties[i];
+            instanceProperties[property] =
+                this.classes[instanceClassId].instances[property][instanceIdx];
+        }
+        // create return object: className: {featureProperties and values}
+        const pickingInfo = {};
+        pickingInfo[featureClass] = instanceProperties;
+        // If this feature has parent(s), recurse on them
+        if (this.inverseHierarchy && this.inverseHierarchy[featureId]) {
+            for (let i = 0; i < this.inverseHierarchy[featureId].length; i++) {
+                const parentID = this.inverseHierarchy[featureId][i];
+                Object.assign(pickingInfo, this.getPickingInfo(parentID));
+            }
+        }
+        return pickingInfo;
+    }
+}
+
+/**
+ * @classdesc
+ * 3D Tiles
+ * [Batch Table Hierarchy Extension](https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/extensions/3DTILES_batch_table_hierarchy)
+ */
+class C3DTBatchTableHierarchyExtension extends C3DTAbstractExtension {
```

Comment:
The idea behind this structure is that one object is in charge of the extension (e.g. `C3DTBatchTableHierarchyExtension` is in charge of the 3DTILES_batch_table_hierarchy extension). Let's name it the extension manager. However, an extension can have data scattered in multiple classes of 3D Tiles (in this case it is only in the batch table but other extensions also have information in tileset, bounding volume, etc.). 
The idea is to have one class per 'piece' of extension (e.g. one class for the piece in the batch table, one class for the piece in the tileset) that is responsible for parsing and possibly doing other thing with this 'piece' of extension. The instance of this class corresponds to the parsed 'piece of extension' and is stored in the corresponding 3D Tiles class and instantiated by the extension manager depending on the calling context (i.e. the 3D Tiles class).

You can see an example implementation [here](https://github.com/jailln/UDV/tree/UDV-temporal/UDV-Core/src/Modules/Temporal/3DTILES_temporal) where we implement an extension of 3D Tiles (not yet in the standard spec) having data in multiple 3D Tiles classes (the entry point is [this file](https://github.com/jailln/UDV/blob/UDV-temporal/UDV-Core/src/Modules/Temporal/3DTILES_temporal/3DTemporalExtension.js)). 

Let me know if my explanation is not clear, I can work on a schema / a sequence diagram.

Reply:
**Common response with this other [comment](https://github.com/iTowns/itowns/pull/1387#discussion_r427359755)**

`C3DTBatchTableHierarchyExtension` is in charge of `C3DT_BTHExtension_BatchTable` ?
Yes, I'd like a diagram to make it clear.

May be, it's possible to simplify, by example :
```js
// Extension class to register in C3DTExtensions
class fooExtension {
    constructor(json) {
        // Parsing json
    }
    // type extension to filter in context
    static get type() { return 'foo' };
}

// in  C3DTExtensions class
C3DTExtensions.parseExtensions(extensionsJSON, type) {
    const parsedExtensions = {};
    for (const extName in extensionsJSON) {
        if (this.isExtensionRegistered(extName)) {
            const extension = this.getExtension(extName);
            if (extension.type == type) {
                parsedExtensions[extName] = new extension(extensionsJSON[extName]);
            }
        }
    }
    return parsedExtensions;
}
```
wdym?

Reply:
In case of the `3DTILES_batch_table_hierarchy` extension, it may work but for extensions in general I think it won't. 3D Tiles extensions can indeed be divided in several parts; each part being stored in a 3D Tiles class. For instance, the [draco point compression extension](https://github.com/CesiumGS/3d-tiles/tree/master/extensions/3DTILES_draco_point_compression) is composed of two parts: one in the `featureTable` class of 3D Tiles and one in the `batchTable` class (see [its JSON schema](https://github.com/CesiumGS/3d-tiles/tree/master/extensions/3DTILES_draco_point_compression/schema)). The implementation therefore follows this idea: the extension is implemented by one class and each extension class is also implemented by one class. When a 3D Tiles class is parsed (e.g. batch table), if an extension part is found it is passed to the corresponding extension which calls the extension part class for the parsing. 

In this PR, `C3DTBatchTableHierarchyExtension` corresponds to the extension and `C3DT_BTHExtension_BatchTable` corresponds to the extension part stored in the batch table.